### DISCUSSION TITLE
I think Ada's subtype is a good feature to reduce program errors very simply, can v add this feature?

### DISCUSSION THREAD

**Discussion by abcxcba** [vlang/v#15637] [Ideas]
About the subtype presentation:

Data types

Ada's type system is not based on a set of predefined [primitive types](https://en.wikipedia.org/wiki/Primitive_types) but allows users to declare their own types. This declaration in turn is not based on the internal representation of the type but on describing the goal which should be achieved. This allows the compiler to determine a suitable memory size for the type, and to check for violations of the type definition at compile time and run time (i.e., range violations, buffer overruns, type consistency, etc.). Ada supports numerical types defined by a range, modulo types, aggregate types (records and arrays), and enumeration types. Access types define a reference to an instance of a specified type; untyped pointers are not permitted. Special types provided by the language are task types and protected types.

For example, a date might be represented as:

```type Day_type   is range    1 ..   31;
type Month_type is range    1 ..   12;
type Year_type  is range 1800 .. 2100;
type Hours is mod 24;
type Weekday is (Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday);

type Date is
   record
     Day   : Day_type;
     Month : Month_type;
     Year  : Year_type;
   end record;

Types can be refined by declaring [subtypes](https://en.wikipedia.org/wiki/Subtyping):

subtype Working_Hours is Hours range 0 .. 12;            -- at most 12 Hours to work a day
subtype Working_Day is Weekday range Monday .. Friday;   -- Days to work

Work_Load: constant array(Working_Day) of Working_Hours  -- implicit type declaration
   := (Friday => 6, Monday => 4, others => 10);           -- lookup table for working hours with initialization
```
Types can have modifiers such as limited, abstract, private etc. Private types can only be accessed and limited types can only be modified or copied within the scope of the package that defines them.[[35]](https://en.wikipedia.org/wiki/Ada_(programming_language)#cite_note-35) Ada 95 adds further features for object-oriented extension of types.

All contents :
[https://en.wikipedia.org/wiki/Ada_(programming_language)#:~:text=Ada%20is%20a%20structured%2C%20statically%20typed%2C%20imperative%2C%20and,tasks%2C%20synchronous%20message%20passing%2C%20protected%20objects%2C%20and%20non-determinism.](url)

**Comment by abcxcba** [comment]
I don't know how to get the code to display like the wiki,  if you can't open the wiki.
You can also see it on this page : https://codedocs.org/what-is/ada-programming-language#Data_types

**Comment by abcxcba** [comment]
Rust’s custom type:
https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html#creating-custom-types-for-validation
```
pub struct Guess {
    value: i32,
}

impl Guess {
    pub fn new(value: i32) -> Guess {
        if value < 1 || value > 100 {
            panic!("Guess value must be between 1 and 100, got {}.", value);
        }

        Guess { value }
    }

    pub fn value(&self) -> i32 {
        self.value
    }
}
```
Compared to Ada, it is both verbose and ugly, but better than nothing.

**Reply by bogen85** [reply]
Yeah. Pascal has a lot of the sub-typing as well, which is where Ada got it from. From what I've seen in Ada the sub-typing is a bit stronger than in Pascal (and C++).

Not sure how these concepts would translate cleanly into V's type system.
